### PR TITLE
Timer

### PR DESCRIPTION
**Pull request by drowsy810301**
- index contest table can auto countdown
- footer clock can auto update through javascript
- re-request server time once a hour to match clock and contest table
- contest table change background when freeze

### CODE REVIEW

File: index/views.py [hunk 1]
```diff
@@ -23,24 +23,27 @@
...
 # Create your views here.
 logger = get_logger()
 def index(request, alert_info='none'):
-
     present = timezone.now()
-    c_runnings = Contest.objects.filter(start_time__lt=present, end_time__gt=present)
-    c_upcomings = Contest.objects.filter(start_time__gt=present)
+    time_threshold = datetime.now() + timedelta(days=1);
+    c_runnings = Contest.objects.filter \
+        (start_time__lt=present, end_time__gt=present,end_time__lt=time_threshold)
```
Comment (nickelcheng) [inline index/views.py]: It makes no sense to ignore running contests which lasts more than one day.
BTW, I think we can ignore homework by add filter ```is_homework=false```.

File: index/static/js/time.js [hunk 2]
```diff
@@ -0,0 +1,112 @@
+/*
```
Comment (nickelcheng) [inline index/static/js/time.js]: I prefer to use ```contestTimer.js``` or something like that. ```time.js``` is ambiguous.

File: index/static/js/time.js [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    var tmp2 = new Array();
+    var tmp3 = new Array();
+    var remainings = new Array();
+    var upcomings = new Array();
+    var flag = 0;
+    try{
+        tmp = document.getElementsByName("end");
+        remainings = document.getElementsByName("remain");
+        tmp2 = document.getElementsByName("contest_id");
+        tmp3 = document.getElementsByName("freeze_time");
```
Comment (nickelcheng) [inline index/static/js/time.js]: I don't think ```tmp``` is a good name. Please use some meaningful variables.

**Comment by bruce3557** [comment]
Other issue, I found the loading speed of contest page is slow.
Any possible solution to solve this?
(Like loading page by ajax and show each section separately)

File: index/static/js/time.js [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+        for(var i=0;i<tmp.length;i++){
+            remaining_time[i] = new Date(tmp[i].textContent).getTime();
+            contest_id[i] = tmp2[i].textContent;
+            freezes[i] = Number(tmp3[i].textContent);
+        }
+    } catch(e){}
```
Comment (bruce3557) [inline index/static/js/time.js]: Please print something here to log exception.

File: index/static/js/time.js [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    tmp = new Array();
+    try{
+        tmp = document.getElementsByName("start");
+        upcomings = document.getElementsByName("upcome");
+        for(var i=0;i<tmp.length;i++)
+            upcoming_time[i] = new Date(tmp[i].textContent).getTime();
+    } catch(e){}
```
Comment (bruce3557) [inline index/static/js/time.js]: Ditto

File: index/static/js/time.js (hunk 4 above)
Comment (drowsy810301) [inline index/static/js/time.js]: How can I print in js?

Comment (bruce3557) [inline index/static/js/time.js]: ```console.log("XXXX")```

Comment (drowsy810301) [inline index/static/js/time.js]: I cannot see anything?

Comment (henryyang42) [inline index/static/js/time.js]: console.log?

**Comment by drowsy810301** [comment]
I found that "getElementsByName" would return empty list if there was no matchup label,so we did not have to use try and catch statement.

**Comment by drowsy810301** [comment]
All commets fixed.

**Comment by nickelcheng** [comment]
+1, @CHIAHSINHOU please test it.

File: index/views.py [hunk 6]
```diff
@@ -23,24 +23,27 @@
...
 # Create your views here.
 logger = get_logger()
 def index(request, alert_info='none'):
-
     present = timezone.now()
-    c_runnings = Contest.objects.filter(start_time__lt=present, end_time__gt=present)
-    c_upcomings = Contest.objects.filter(start_time__gt=present)
+    time_threshold = datetime.now() + timedelta(days=1);
+    c_runnings = Contest.objects.filter \
+        (start_time__lt=present, end_time__gt=present, is_homework=False)
```
Comment (bruce3557) [inline index/views.py]: not sure why double ```_```  here

Comment (nickelcheng) [inline index/views.py]: It's django build in filter.

Comment (bruce3557) [inline index/views.py]: ok

**Comment by bruce3557** [comment]
+1, going to merge after testing

**Comment by CHIAHSINHOU** [comment]
go to merge

**Pull request merged**